cpp/src/arrow/util/async_util.cc

```diff
@@ -168,6 +146,25 @@ class AsyncTaskSchedulerImpl : public AsyncTaskScheduler {
     if (IsAborted()) {
       return false;
     }
+#ifdef ARROW_WITH_OPENTELEMETRY
+    // Wrap the task to propagate a parent tracing span to it
+    struct WrapperTask : public Task {
+      WrapperTask(
+          std::unique_ptr<Task> target,
+          opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> parent_span)
+          : target(std::move(target)), parent_span(std::move(parent_span)) {}
+      Result<Future<>> operator()() override {
+        auto scope = arrow::internal::tracing::GetTracer()->WithActiveSpan(parent_span);
+        return (*target)();
+      }
+      int cost() const override { return target->cost(); }
+      std::string_view name() const override { return target->name(); }
+      std::unique_ptr<Task> target;
+      opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> parent_span;
+    };
+    task = std::make_unique<WrapperTask>(
+        std::move(task), arrow::internal::tracing::GetTracer()->GetCurrentSpan());
```

Comment:
Note that I was intentionally not using `GetCurrentSpan` as the parent.  We are submitting a new task here.  I would like that new task to appear as its own span and not as a child of the thing that submitted it.  For example, in Acero, in most plans, the source node is the one that schedules all the tasks.  If we use `GetCurrentSpan` here then it will look like all the work being done is a part of the source node.